artiq/gateware/rtio/phy/servo.py

```diff
@@ -150,14 +149,25 @@ def __init__(self, w, servo):
                     [_.clip.eq(0) for _ in servo.iir.ctrl]
                 )
         ]
+
+        o_data_w = len(self.rtlink.o.data)
+        i_data_w = len(self.rtlink.i.data)
+
+        i_data = Signal(i_data_w)
+        i_data_sign = Signal()
+
         self.comb += [
                 self.rtlink.i.stb.eq(read),
-                self.rtlink.i.data.eq(
-                    Mux(read_state,
-                        Mux(read_config,
-                            status,
-                            m_state.dat_r[w.state - w.coeff:]),
-                        Mux(read_high,
-                            m_coeff.dat_r[w.coeff:],
-                            m_coeff.dat_r[:w.coeff])))
+                self.rtlink.i.data.eq(i_data),
+                i_data[:w.coeff].eq(Mux(read_state,
+                                        Mux(read_config,
+                                            status,
+                                            m_state.dat_r[w.state - w.coeff:]),
+                                        Mux(read_high,
+                                            m_coeff.dat_r[w.coeff:],
+                                            m_coeff.dat_r[:w.coeff]))),
+                i_data_sign.eq(i_data[w.coeff-1]),
+                i_data[w.coeff:].eq(Mux(i_data_sign,
+                                        (1 << i_data_w-w.coeff)-1,
+                                        0))
```

Comment:
Is it correct to sign extend for all reads? Are the unsigned ones all small enough to not have a sign bit there?
Could you add a comment about what you are doing here?
Please watch the style with the spaces.
And I think just doing the following should be all you need:

```python
def sign_extend(x, width):
    return x[width:].eq(Repeat(x[width - 1], len(x) - width))

# sign extend the read port
# TODO: note about unsigned coefficients
sign_extend(self.rtlink.i.data, w.coeff)
```

Reply:
> Is it correct to sign extend for all reads? Are the unsigned ones all small enough to not have a sign bit there?

Yes, I believe that with the current field widths that's correct (but correct me if I'm wrong). If, for example, the coefficient width were reduced to 16, this would no longer hold. So, we should probably add a comment about this in the code. Maybe also an assertion?

Briefly:
- everything in the state memory (`x0`, `x1`, `y1`) is treated as a signed `w.coeff`-bit number (internally wider, but RTIO interface only reads the MSB)
- the IIR coefficients (`a0`, `b1`, `b0`) and `offset` are all `w.coeff`-bit signed
- the DDS ftw halves + pow are 16-bit unsigned
- `cfg` is 16-bits unsigned (8-bits of ADC channel select and 8-bit delay)
- `status` is currently 16-bit (lower 8 bits are [`start`, `done`, pad], upper 8 bits are `clip` indicators).

The fuzzing test above verifies that after this PR all state entries are read back correctly. Additionally I've briefly checked the ADC reading on an experiment.

Reply:
> Please watch the style with the spaces.

Which bit did you have in mind? IIRC this was spaced the way flake8 requested.

> And I think just doing the following should be all you need:

Yes, that's a neater way of expressing the logic. We can do it that way if you prefer.